### PR TITLE
react-native link (aka rnpm) support for Windows

### PR DESCRIPTION
**Pull request by GantMan**
Seeing as [Windows is a supported platform](https://github.com/facebook/react-native/blob/72157cf99164d00c7a14b6b9ca51b406080b5265/packager/defaults.js#L22) until platforms can better manager their own CLI and packager needs.

### Motivation
Linking 3rd party libraries should be supported first, because then I'd like to do a follow up PR with @grabbou to identify how we can effectively move RNPM functionality out of react-native core and eventually housed in each external platform's repo.  The goal would be working with @cpojer and hopefully @andrewimm to help keep external platform needs in their respective repos, for rnpm/packager _et al._  Seeing as this is a major discussion point, I've made this PR first.  Making small steps towards this goal, seems to be the approved methodology from all.   

Additionally, I have a merged PR that makes an excellent place for documenting the CLI when it advances, as preparation for this initiative: ([previous PR](https://github.com/facebook/react-native/pull/11097))

### About this PR
We're no stranger to rnpm/react-native link:  This brings that known experience to this platform.

* Here is my example: [Project ready for linking](https://github.com/GantMan/rnpm_windows_example)
* Running `react-native link`
screen shot 2016-12-02 at 2 07 47 pm
* Running `react-native link` again
screen shot 2016-12-02 at 2 08 12 pm
The resulting code is available in this branch: [Example project linked](https://github.com/GantMan/rnpm_windows_example/tree/linked)
* Running `react-native unlink react-native-win-gif` twice
screen shot 2016-12-03 at 11 05 57 am
* The resulting code is available in this branch: [Example project unlinked](https://github.com/GantMan/rnpm_windows_example/tree/unlinked)

### Code style
This code is meant to follow the exact style of the previous structures and commands.  For anyone familiar with the existing CLI framework, this should fit accordingly.  There were some new needs for revoke patches that might be worth eventually merging into a shared codebase between Android and C#.

### CODE REVIEW

**Comment by cpojer** [comment]
@GantMan I just wanna make sure that we are on the same page wrt ownership. The JavaScript Tools team at FB London owns React Native Packager. `rnpm` is a community contribution that was merged into react-native and it isn't part of packager. So I'm happy to work with you on all the RNP related things but for `rnpm` I'm afraid I can't really help that much. Just thought I clear this up in case there is any confusion here. We don't use `rnpm` at Facebook, so our React Stack team is the one that should be looking at this PR specifically (cc @mkonicek).

**Comment by GantMan** [comment]
Hey @cpojer :)  Yes, sorry to mention you on this.  I'm aware it's unrelated to your direct tasks on packager.  I mentioned you because I'm speaking on the concepts we had mentioned for platforms, since we're all having to think about how to best handle multiple platforms (across RNPM and RNP).  In this sense, I would appreciate enjoy your expertise in theory.   As far as the code, I expect @grabbou and @rozele to chime in.  I've already discussed this PR with them.

**Comment by GantMan** [comment]
the failing test seems unrelated to the changes.  Can some one corroborate?  Looks like CI is failing a lot for even doc update PRs?

**Comment by grabbou** [comment]
Thanks! I'll follow up on this tomorrow. Let's move this discussion onto Slack for faster iteration. Anyone who wants to be in the loop - please let me know. There's going to be a bit of work and by using a synchronous way of communicating we can speed it up!

File: local-cli/link/windows/patches/makeProjectPatch.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+module.exports = function makeProjectPatch(projectInsert, projectName = '') {
+  return {
+    pattern: '<ProjectReference Include="..\\..\\node_modules\\react-native-windows\\ReactWindows\\ReactNative\\ReactNative.csproj">',
```
Comment (rozele) [inline local-cli/link/windows/patches/makeProjectPatch.js]: >ProjectReference [](start = 15, length = 16)

I don't think this works.  This is typically followed by:
```
      <Project>{c7673ad5-e3aa-468c-a5fd-fa38154e205c}</Project>
      <Name>ReactNative</Name>
    </ProjectReference>
```
And the full project reference should come after the close </ProjectReference>

File: local-cli/link/windows/patches/makePackagePatch.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+const applyParams = require('./applyParams');
+
+module.exports = function makePackagePatch(packageInstance, params, prefix) {
+  const processedInstance = applyParams(packageInstance, params, prefix);
+
+  return {
+    pattern: 'new MainReactPackage()',
```
Comment (rozele) [inline local-cli/link/windows/patches/makePackagePatch.js]: Is it possible to capture regex groups and re-apply them in the patch?  If so, you could capture the whitespace for the `(\s*)new MainReactPackage()` and apply it to the patch so there's matching whitespace.  Thinking something like:
```
{
    pattern: '(\s*)new MainReactPackage()',
    patch: ",\n$1' + processedInstance
}
```

File: local-cli/link/windows/patches/makeProjectPatch.js (hunk 1 above)
Comment (rozele) [inline local-cli/link/windows/patches/makeProjectPatch.js]: Nevermind, you use the `flip` parameter to make sure the patch comes before the ProjectReference.

---
In reply to: [90884760](https://github.com/facebook/react-native/pull/11282#discussion_r90884760) [](ancestors = 90884760)

File: local-cli/link/windows/patches/makePackagePatch.js (hunk 2 above)
Comment (rozele) [inline local-cli/link/windows/patches/makePackagePatch.js]: Unless you use the `flip` parameter here, then you can add the comma after `processedInstance` to guarantee it's always correct.

---
In reply to: [90885370](https://github.com/facebook/react-native/pull/11282#discussion_r90885370) [](ancestors = 90885370)

File: local-cli/core/config/windows/generateGUID.js [hunk 3]
```diff
@@ -0,0 +1,10 @@
+const s4 = () => {
+  return Math.floor((1 + Math.random()) * 0x10000)
+    .toString(16)
+    .substring(1);
+}
+
+module.exports = function generateGUID() {
```
Comment (rozele) [inline local-cli/core/config/windows/generateGUID.js]: >generateGUID [](start = 26, length = 12)

Nice, I should use this and remove the dependency on `uuid` in react-native-windows CLI.

File: local-cli/link/windows/patches/makeProjectPatch.js (hunk 1 above)
Comment (GantMan) [inline local-cli/link/windows/patches/makeProjectPatch.js]: yup flip works 👍

File: local-cli/core/config/windows/index.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      <Name>${projectName}</Name>
+    </ProjectReference>
+    `;
+
+  return {
+    sourceDir,
+    packageUsingPath,
+    packageInstance,
+    solutionInsert,
+    projectInsert,
```
Comment (GantMan) [inline local-cli/core/config/windows/index.js]: I can see about moving this down to patch level.

File: local-cli/link/windows/patches/makePackagePatch.js (hunk 2 above)
Comment (GantMan) [inline local-cli/link/windows/patches/makePackagePatch.js]: will give this a try and let you know.

File: local-cli/core/config/windows/findNamespace.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+ * by searching for its declaration in all C# files present in the folder
+ *
+ * @param {String} folder Folder to find C# files
+ */
+module.exports = function getNamespace(folder) {
+  const files = glob.sync('**/*.cs', { cwd: folder });
+
+  const packages = files
+    .map(filePath => fs.readFileSync(path.join(folder, filePath), 'utf8'))
+    .map(file => file.match(/namespace (.*)\s*/))
```
Comment (GantMan) [inline local-cli/core/config/windows/findNamespace.js]: Will check into this and report back.

Comment (GantMan) [inline local-cli/core/config/windows/findNamespace.js]: I punched up the Regex so it finds the first namespace that is followed by `IReactPackage`

File: local-cli/core/config/windows/index.js (hunk 4 above)
Comment (GantMan) [inline local-cli/core/config/windows/index.js]: moved down to patches.

**Comment by grabbou** [comment]
Thanks @GantMan for the contribution, as I already said via Slack! It's awesome :) Let me just go through the diff and leave some inline comments.

File: local-cli/link/link.js [hunk 6]
```diff
@@ -58,6 +60,33 @@ const linkDependencyAndroid = (androidProject, dependency) => {
   });
 };
 
+const linkDependencyWindows = (windowsProject, dependency) => {
+
```
Comment (grabbou) [inline local-cli/link/link.js]: nit: empty line

**Comment by grabbou** [comment]
Going to merge it, impressive work!

@rozele how do you feel about Windows-specific code finally landing in RN code? :trollface: 

Seriously speaking, here's the plan:
1) Next PR - remove `windows` related code from this repo and put it inside `react-native-windows/local-cli/core/windows` so it still mirrors this repo. We merged this one for now to test it faster, get it out faster.
2) Next PR - make all the getters on `rnpm config` lazy, so that they don't run when you don't need them. That should significantly speed up the CLI, because who needs the config to run when there's `start` command running?

**Comment by grabbou** [comment]
@facebook-github-bot shipit

**Pull request closed without merging**